sdk/java-sdk-protobuf/src/main/scala/kalix/javasdk/impl/MetadataImpl.scala

```diff
@@ -194,6 +196,13 @@ private[kalix] class MetadataImpl(val entries: Seq[MetadataEntry]) extends Metad
 
   override def clearTime(): MetadataImpl = remove(MetadataImpl.CeTime)
 
+  override def withHttpResponseCode(code: StatusCode): MetadataImpl = {
+    if (code.isInstanceOf[ErrorCode])
+      throw new RuntimeException("Invalid HTTP status code: " + code + ". Use .error method for replying with errors.")
+
+    set("_kalix-http-code", code.value.toString)
+  }
```

Comment:
One thing to consider, if given direct official access to this: redirects for example requires a `Location` header where the redirect points to, but now a user could say `301 Moved permanently` without such a header resulting in an invalid HTTP response. There are also a bunch of codes that are more protocol level and should never be emitted by a user endpoint (100, 103 for example).

That might be fine (with a "if it hurts, don't do it"-mentality) but could also mean we should rather have higher level APIs when we make this more officially supported.

Reply:
Yep, definitely good points.

I was doing some experiments today but I'm not so confident we know exactly where we want to go, mainly for code-first sdk (depending on spring classes vs having our own annotations). Wrote some more [here](https://github.com/lightbend/kalix-jvm-sdk/issues/1851#issuecomment-1875576570).